codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoderTest.java

```diff
@@ -174,17 +174,54 @@ public void dataReadAfterGoAwayShouldApplyFlowControl() throws Exception {
         int processedBytes = data.readableBytes() + padding;
         mockFlowControl(processedBytes);
         try {
-            decode().onDataRead(ctx, STREAM_ID, data, 10, true);
+            decode().onDataRead(ctx, STREAM_ID, data, padding, true);
             verify(localFlow).receiveFlowControlledFrame(eq(ctx), eq(stream), eq(data), eq(padding), eq(true));
             verify(localFlow).consumeBytes(eq(ctx), eq(stream), eq(processedBytes));
 
-            // Verify that the event was absorbed and not propagated to the oberver.
+            // Verify that the event was absorbed and not propagated to the observer.
             verify(listener, never()).onDataRead(eq(ctx), anyInt(), any(ByteBuf.class), anyInt(), anyBoolean());
         } finally {
             data.release();
         }
     }
 
+    @Test
+    public void emptyDataFrameWithPaddingShouldApplyFlowControl() throws Exception {
+        final ByteBuf data = EMPTY_BUFFER;
+        int padding = 10;
+        int processedBytes = data.readableBytes() + padding;
+        mockFlowControl(processedBytes);
+        try {
+            decode().onDataRead(ctx, STREAM_ID, data, padding, true);
+            verify(localFlow).receiveFlowControlledFrame(eq(ctx), eq(stream), eq(data), eq(padding), eq(true));
+            verify(localFlow).consumeBytes(eq(ctx), eq(stream), eq(processedBytes));
+
+            // Verify that the empty data event was propagated to the observer.
+            verify(listener).onDataRead(eq(ctx), eq(STREAM_ID), eq(data), eq(padding), eq(true));
+        } finally {
+            data.release();
+        }
+    }
+
+    @Test
+    public void emptyDataFrameWithoutPaddingShouldNotApplyFlowControl() throws Exception {
```

Comment:
Lets instead change this to make sure that it does get called (especially if `endOfStream` is set).  If this change duplicates another unit test (then remove it) or if it can be combined with another unit test (then that works too).